### ISSUE TITLE
possible-bug-revise-logic-for-repetition-penalty-in-generate-loop

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/cerebros-core-algorithm-alpha#294]
# Issue Described: Repetition Penalty (Incorrect Implementation)

- Code: logits[token_id] / repetition_penalty
- Standard Definition:  (if logit != 0 divide by penalty else multiply by penalty)

#  AI suggested correction:

```python
if repetition_penalty is not None and repetition_penalty != 1.0:
    vocab_size = tf.shape(logits)[0]

    # int32 tensor of unique token ids
    token_ids = tf.unique(tf.cast(current_tokens, tf.int32)).y
    token_ids = tf.boolean_mask(token_ids, token_ids < vocab_size)

    # gather logits for those tokens
    selected_logits = tf.gather(logits, token_ids)

    # apply penalty: divide positive, multiply negative
    pos = selected_logits > 0
    new_selected_logits = tf.where(
        pos,
        selected_logits / repetition_penalty,
        selected_logits * repetition_penalty,
    )

    # scatter back in one shot
    logits = tf.tensor_scatter_nd_update(
        logits,
        indices=tf.expand_dims(token_ids, axis=1),
        updates=new_selected_logits,
    )
```
